components/scream/src/share/property_checks/mass_and_energy_column_conservation_check.cpp

```diff
@@ -0,0 +1,331 @@
+  Kokkos::parallel_for(policy, KOKKOS_LAMBDA (const KT::MemberType& team) {
+
+    const int i = team.league_rank();
+
+    const auto pseudo_density_i = ekat::subview(pseudo_density, i);
+    const auto T_mid_i          = ekat::subview(T_mid, i);
+    const auto horiz_winds_i    = ekat::subview(horiz_winds, i);
+    const auto qv_i             = ekat::subview(qv, i);
+    const auto qc_i             = ekat::subview(qc, i);
+    const auto qi_i             = ekat::subview(qi, i);
+    const auto qr_i             = ekat::subview(qr, i);
+
+    // Calculate total mass
+    const Real tm = compute_total_mass_on_column(team, pseudo_density_i, qv_i, qc_i, qi_i, qr_i);
+    const Real previous_tm = mass(i);
+
+    // Calculate expected total mass. Here, dt should be set to the timestep of the 
+    // subcycle for the process that called this check. This effectively scales the boundary
+    // fluxes by 1/num_subcycles (dt = model_dt/num_subcycles) so that we only include
+    // the expected change after one substep (not a full timestep).
+    const Real tm_exp = previous_tm +
+                        compute_mass_boundary_flux_on_column(vapor_flux(i), water_flux(i))*dt;
+
+    // Calculate relative error of total mass
+    const Real err_m = std::abs(tm-tm_exp);
+    rel_err_mass(i) = err_m/previous_tm;
+
+    // Calculate total energy
+    const Real te = compute_total_energy_on_column(team, pseudo_density_i, T_mid_i, horiz_winds_i,
+                                                   qv_i, qc_i, qr_i, ps(i), phis(i));
+    const Real previous_te = energy(i);
+
+    // Calculate expected total energy. See the comment above for an explanation of dt. 
+    const Real te_exp = previous_te +
+                        compute_energy_boundary_flux_on_column(vapor_flux(i), water_flux(i), ice_flux(i), heat_flux(i))*dt;
+
+    // Calculate relative error of total energy
+    const Real err_e = std::abs(te-te_exp);
+    rel_err_energy(i) = err_e/previous_te;
+  });
+
+  // Use Kokkos::MaxLoc to find the largest error for both mass and energy
+  // and test that they are below the tolerance.
+  using maxloc_t = Kokkos::MaxLoc<Real, int>;
+  using maxloc_value_t = typename maxloc_t::value_type;
+  maxloc_value_t maxloc_mass;
+  maxloc_value_t maxloc_energy;
+
+  Kokkos::parallel_reduce("mass_reduce", m_num_cols, KOKKOS_LAMBDA(int i, maxloc_value_t& result) {
+    if (rel_err_mass(i) > result.val) {
+      result.val = rel_err_mass(i);
+      result.loc = i;
+    }
+  }, maxloc_t(maxloc_mass));
+
+  Kokkos::parallel_reduce("energy_reduce", m_num_cols, KOKKOS_LAMBDA(int i, maxloc_value_t& result) {
+    if (rel_err_energy(i) > result.val) {
+      result.val = rel_err_energy(i);
+      result.loc = i;
+    }
+  }, maxloc_t(maxloc_energy));
+
+  PropertyCheck::ResultAndMsg res_and_msg;
+  if (maxloc_mass.val < m_tol && maxloc_energy.val < m_tol) {
+    res_and_msg.result = CheckResult::Pass;
```

Comment:
If the check passed, do we really want to spend time creating a message that will never be printed anyways? Imho, we can do an early return here...